solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/outbound/JCSMPOutboundMessageHandler.java

```diff
@@ -96,19 +101,43 @@ public void handleMessage(Message<?> message) throws MessagingException {
 					String.format("Unable to parse header %s", SolaceBinderHeaders.CONFIRM_CORRELATION), e);
 		}
 
-		XMLMessage xmlMessage = xmlMessageMapper.map(message, properties.getExtension().getHeaderExclusions(),
-				properties.getExtension().isNonserializableHeaderConvertToString());
-		correlationKey.setRawMessage(xmlMessage);
-		xmlMessage.setCorrelationKey(correlationKey);
+		List<XMLMessage> smfMessages;
 
-		LOGGER.debug("Publishing message to destination [ {}:{} ]",
-				targetDestination instanceof Topic ? "TOPIC" : "QUEUE", targetDestination);
+		if (message.getHeaders().containsKey(SolaceBinderHeaders.BATCHED_HEADERS)) {
+			LOGGER.debug("Detected header {}, handling as batched message (Message<List<?>>) <message handler ID: {}>",
+					SolaceBinderHeaders.BATCHED_HEADERS, id);
+			smfMessages = xmlMessageMapper.mapBatchMessage(
+					message,
+					properties.getExtension().getHeaderExclusions(),
+					properties.getExtension().isNonserializableHeaderConvertToString());
+		} else {
+			smfMessages = List.of(xmlMessageMapper.map(
+					message,
+					properties.getExtension().getHeaderExclusions(),
+					properties.getExtension().isNonserializableHeaderConvertToString()));
+		}
 
 		try {
-			producer.send(xmlMessage, targetDestination);
+			for (int i = 0; i < smfMessages.size(); i++) {
+				XMLMessage smfMessage = smfMessages.get(i);
+				correlationKey.setRawMessage(smfMessage);
+				smfMessage.setCorrelationKey(correlationKey);
```

Comment:
In batch mode, would this resolve the correlationKey N times? Should the correlation key be resolved only once, either successfully or with failure, when the whole batch has been processed?

Reply:
ah, that's a good catch. That's a bug. Right now, it will resolve the correlation key when the first message is received instead of the last. Let me fix that...

Reply:
Fixed

Reply:
The batched message is a work item. The intention of the corresponding `CorrelationData` at the root of the batched message is to say whether or not the work item as a whole passes or fails.